falcon/asgi/ws.py

```diff
@@ -0,0 +1,340 @@
+        #   errors.
+        if not isinstance(payload, str):
+            raise TypeError('payload must be a string')
+
+        await self._send({
+            'type': 'websocket.send',
+            'text': payload,
+        })
+
+    async def send_data(self, payload: Union[bytes, bytearray, memoryview]) -> None:
+        # NOTE(kgriffs): We have to check ourselves because some ASGI
+        #   servers are not very strict which can lead to hard-to-debug
+        #   errors.
+        if not isinstance(payload, (bytes, bytearray, memoryview)):
+            raise TypeError('payload must be a byte string')
+
+        self._require_accepted()
+
+        # NOTE(kgriffs): Assume bytes, bytearray, or memoryview
+        await self._send({
+            'type': 'websocket.send',
+            'bytes': bytes(payload),
+        })
+
+    async def receive_text(self) -> str:
+        self._require_accepted()
+
+        event = await self._receive()
+
+        # PERF(kgriffs): When we normally expect the key to be
+        #   present, this pattern is faster than get()
+        try:
+            text = event['text']
+        except KeyError:
+            text = None
+
+        # NOTE(kgriffs): Even if the key is present, it may be None
+        if text is None:
+            raise errors.MessagePayloadError('Expected TEXT payload but got BINARY instead')
+
+        return text
+
+    async def receive_data(self) -> bytes:
+        self._require_accepted()
+
+        event = await self._receive()
+
+        # PERF(kgriffs): When we normally expect the key to be
+        #   present, EAFP is faster than get()
+        try:
+            data = event['bytes']
+        except KeyError:
+            data = None
+
+        # NOTE(kgriffs): Even if the key is present, it may be None
+        if data is None:
+            raise errors.MessagePayloadError('Expected BINARY payload but got TEXT instead')
+
+        return data
+
+    async def receive_media(self) -> object:
+        self._require_accepted()
+
+        event = await self._receive()
+
+        # NOTE(kgriffs): Most likely case is going to be JSON via text
+        #   payload, so try that first.
+        text = event.get('text')
+        if text is not None:
+            return self._mh_text_deserialize(text)
+
+        # PERF(kgriffs): At this point there better be a 'bytes' key, so
+        #   use EAFP this time.
+        try:
+            data = event['bytes']
+        except KeyError:
+            raise errors.MessagePayloadError(
+                'Expected TEXT or BINARY payload but got None instead'
+            )
+
+        return self._mh_bin_deserialize(data)
+
+    async def _send(self, msg: dict):
+        try:
+            await self._asgi_send(msg)
+        except Exception as ex:
+            # NOTE(kgriffs): likely uvicorn (or some other project using the
+            #   the the "websockets" Python library. Daphne, at least, doesn
```

Comment:
A generous amount of `the`s in a row :slightly_smiling_face:

Reply:
reworded